sim/midas/src/main/scala/midas/core/FPGATop.scala

```diff
@@ -117,6 +118,70 @@ case class CPUManagedAXI4Params(
   )
 }
 
+class QSFPBundle(qsfpBitWidth: Int) extends Bundle {
+  val channel_up = Input(Bool())
+  val tx = Decoupled(UInt(qsfpBitWidth.W))
+  val rx = Flipped(Decoupled(UInt(qsfpBitWidth.W)))
+}
+
+object QSFPBundle {
+  def apply(qsfpBitWidth: Int)(implicit p: Parameters): QSFPBundle = {
+    new QSFPBundle(qsfpBitWidth)
+  }
+}
+
+class SerialIO(val w: Int) extends Bundle {
+  val in = Flipped(Decoupled(UInt(w.W)))
+  val out = Decoupled(UInt(w.W))
+
+  def flipConnect(other: SerialIO) {
+    in <> other.out
+    other.in <> out
+  }
+}
+
+class SerialWidthAdapter(narrowW: Int, wideW: Int) extends Module {
+  require(wideW > narrowW)
+  require(wideW % narrowW == 0)
+  val io = IO(new Bundle {
+    val narrow = new SerialIO(narrowW)
+    val wide = new SerialIO(wideW)
+  })
+
+  val beats = wideW / narrowW
+  val narrow_beats = RegInit(0.U(log2Ceil(beats).W))
+  val narrow_last_beat = narrow_beats === (beats-1).U
+  val narrow_data = Reg(Vec(beats-1, UInt(narrowW.W)))
+
+  val wide_beats = RegInit(0.U(log2Ceil(beats).W))
+  val wide_last_beat = wide_beats === (beats-1).U
+
+  io.narrow.in.ready := Mux(narrow_last_beat, io.wide.out.ready, true.B)
+  when (io.narrow.in.fire()) {
+    narrow_beats := Mux(narrow_last_beat, 0.U, narrow_beats + 1.U)
+    when (!narrow_last_beat) { narrow_data(narrow_beats) := io.narrow.in.bits }
+  }
+  io.wide.out.valid := narrow_last_beat && io.narrow.in.valid
+  io.wide.out.bits := Cat(io.narrow.in.bits, narrow_data.asUInt)
+
+  io.narrow.out.valid := io.wide.in.valid
+  io.narrow.out.bits := io.wide.in.bits.asTypeOf(Vec(beats, UInt(narrowW.W)))(wide_beats)
+  when (io.narrow.out.fire()) {
+    wide_beats := Mux(wide_last_beat, 0.U, wide_beats + 1.U)
+  }
+  io.wide.in.ready := wide_last_beat && io.narrow.out.ready
+}
+
+object FPGATopLogger {
+  def logInfo(format: String, args: Bits*)(implicit p: Parameters) {
+    val loginfo_cycles = RegInit(0.U(64.W))
+    loginfo_cycles := loginfo_cycles + 1.U
+
+    printf("cy: %d, ", loginfo_cycles)
+    printf(Printable.pack(format, args:_*))
+  }
+}
```

Comment:
Since these are more general utilities can you move them to a new file(s) and label where they came from (in the case of the SerialIO items).